waiter/integration/waiter/websocket_integration_test.clj

```diff
@@ -153,6 +160,62 @@
         (finally
           (delete-service waiter-url waiter-headers))))))
 
+(deftest ^:parallel ^:integration-slow test-fail-to-stream-large-responses
+  (testing-using-waiter-url
+    (let [^WebSocketClient websocket-client (websocket-client-factory)
+          waiter-settings (waiter-settings waiter-url)
+          {:keys [ws-max-binary-message-size ws-max-text-message-size]} (:websocket-config waiter-settings)
+          ws-max-binary-message-size' (+ 2048 ws-max-binary-message-size)
+          ws-max-text-message-size' (+ 2048 ws-max-text-message-size)
+          auth-cookie-value (auth-cookie waiter-url)
+          waiter-headers (-> (kitchen-request-headers)
+                             (assoc :x-waiter-mem 1024
+                                    :x-waiter-metric-group "test-ws-support"
+                                    :x-waiter-name (rand-name))
+                             (update :x-waiter-cmd
+                                     (fn [cmd] (str cmd ;; on-the-fly doesn't support x-waiter-env
+                                                    " --mem 1024M"
+                                                    " --ws-max-binary-message-size " ws-max-binary-message-size'
+                                                    " --ws-max-text-message-size " ws-max-text-message-size'))))
+          middleware (fn middleware [_ ^UpgradeRequest request]
+                       (websocket/add-headers-to-upgrade-request! request waiter-headers)
+                       (add-auth-cookie request auth-cookie-value))]
+      (update-max-message-sizes websocket-client ws-max-binary-message-size' ws-max-text-message-size')
+      (is auth-cookie-value)
+      (try
+
+        ;; avoiding testing chars as it causes kitchen to run out of memory
+        (testing "large binary response"
+          (let [response-promise (promise)
+                backend-data-promise (promise)
+                ctrl-data-promise (promise)]
+            (ws-client/connect!
+              websocket-client
+              (ws-url waiter-url "/websocket-byte-stream")
+              (fn [{:keys [in out ctrl]}]
+                (async/go
+                  (async/>! out "first-message")
+                  (async/<! in) ;; kitchen message
+                  (async/<! in) ;; hello response
+                  (async/>! out (str "bytes-" (+ ws-max-text-message-size 1024)))
+                  (let [backend-response (async/<! in)]
+                    (deliver backend-data-promise backend-response))
+                  (async/close! out)
+                  (let [ctrl-data (async/<! ctrl)]
+                    (deliver ctrl-data-promise ctrl-data))
+                  (deliver response-promise :done)))
+              {:middleware middleware})
+
+            (is (= :done (deref response-promise (-> 3 t/minutes t/in-millis) :timed-out)))
```

Comment:
If everyone's on board, let's do that as a separate refactor then? There are other websocket integration tests that are also doing async reads.